Source/WebCore/dom/TrustedTypeUtils.cpp

```diff
@@ -0,0 +1,162 @@
+        auto trustedTypesFactory = WindowOrWorkerGlobalScopeTrustedTypes::trustedTypes(*workerGlobalScope);
+        protectedPolicy = trustedTypesFactory->defaultPolicy();
+    }
+
+    if (!protectedPolicy)
+        return nullptr;
+
+    VM& vm = scriptExecutionContext.vm();
+
+    auto jsExpectedType = JSC::jsString(vm, trustedTypeToString(expectedType));
+    JSC::Strong<JSC::Unknown> strongExpectedType(vm, jsExpectedType);
+    auto jsSink = JSC::jsString(vm, sink);
+    JSC::Strong<JSC::Unknown> strongSink(vm, jsSink);
+    FixedVector<JSC::Strong<JSC::Unknown>> arguments({ strongExpectedType, strongSink });
+    auto policyValueHolder = protectedPolicy->getPolicyValue(expectedType, input, WTFMove(arguments), false);
+    if (policyValueHolder.hasException())
+        return { policyValueHolder.releaseException() };
+
+    auto policyValue = policyValueHolder.releaseReturnValue();
+    if (policyValue.isNull())
+        return nullptr;
+
+    if (expectedType == TrustedType::TrustedHTML)
+        return { TrustedHTML::create(policyValue) };
+    if (expectedType == TrustedType::TrustedScript)
+        return { TrustedScript::create(policyValue) };
+    if (expectedType == TrustedType::TrustedScriptURL)
+        return { TrustedScriptURL::create(policyValue) };
+
+    ASSERT_NOT_REACHED();
+    return nullptr;
+}
+
+// https://w3c.github.io/trusted-types/dist/spec/#get-trusted-type-compliant-string-algorithm
+ExceptionOr<String> getTrustedTypeCompliantString(TrustedType expectedType, ScriptExecutionContext& scriptExecutionContext, const String& input, const String& sink)
+{
+    String stringValue(input);
+
+    if (!scriptExecutionContext.settingsValues().trustedTypesEnabled)
+        return stringValue;
+
+    CheckedPtr contentSecurityPolicy = scriptExecutionContext.checkedContentSecurityPolicy();
+
+    auto requireTrustedTypes = contentSecurityPolicy
+        ? contentSecurityPolicy->requireTrustedTypesForSinkGroup("script"_s)
+        : false;
+
+    if (!requireTrustedTypes)
+        return stringValue;
+
+    auto convertedInput = processValueWithDefaultPolicy(scriptExecutionContext, expectedType, stringValue, sink);
+    if (std::holds_alternative<Exception>(convertedInput))
+        return WTFMove(std::get<Exception>(convertedInput));
+
+    if (!std::holds_alternative<std::nullptr_t>(convertedInput)) {
+        stringValue = std::visit(TrustedTypeVisitor { }, convertedInput);
+        if (stringValue.isNull())
+            convertedInput = nullptr;
+    }
+
+    if (std::holds_alternative<std::nullptr_t>(convertedInput)) {
+        auto allowMissingTrustedTypes = contentSecurityPolicy->allowMissingTrustedTypesForSinkGroup(trustedTypeToString(expectedType), sink, "script"_s, stringValue);
+
+        if (!allowMissingTrustedTypes)
+            return Exception { ExceptionCode::TypeError, makeString("This assignment requires a ", trustedTypeToString(expectedType)) };
```

Comment:
Seems like we might want the message here to be more human readable.
> This assignment requires a trusted HTML string.

rather than

> This assignment requires a TrustedHTML

Reply:
I have a follow up issue to go through the error messages to ensure they're as helpful as possible there's some rather specific cases where we can be more descriptive. I'm happy to address that then, or can do it now if you'd prefer.